Scheduler interface with `choose`

yihozhang opened the pull request:
Closes #431

Alex-Fischman commented on src/scheduler.rs:
```diff
@@ -0,0 +1,428 @@
+    }
+
+    pub fn choose(&mut self, idx: usize) {
+        self.chosen.set(idx, true);
```
Can we avoid the linear retain if we get rid of the bitvec and just immediately do the vec removal and table insertion here? We would want to do something with the idx so that we could use `swap_remove`.

Alex-Fischman commented on src/scheduler.rs:
```diff
@@ -0,0 +1,428 @@
+    }
+}
+
+fn collect_rules<'a>(
```
Can we factor out this function and `collect_rule_ids` in `lib.rs`? They do the same thing.

yihozhang replied:
It is doing a slightly different thing since it's collecting ResolvedCoreRules instead of RuleIds. I put the definition inside `step_rule_with_scheduler` to be consistent with `step_rules`

ezrosent reviewed (commented):
Seems reasonable. Thanks for explaining things offline as well.

ezrosent commented on Cargo.toml:
```diff
@@ -56,12 +56,17 @@ rustc-hash = "1.1"
+core-relations = { git = "https://github.com/egraphs-good/egglog-backend.git", rev = "32c630" }
+egglog-bridge = { git = "https://github.com/egraphs-good/egglog-backend.git", rev = "32c630" }
+numeric-id = { git = "https://github.com/egraphs-good/egglog-backend.git", rev = "32c630" }
+# core-relations = { path = "../egglog-backend/core-relations" }
```
clean these up before submitting?

Alex-Fischman replied:
Honestly I'm okay with the comments, I have the same directory structure and it's annoying to switch manually.

yihozhang commented:
@Alex-Fischman @ezrosent I have refactored to use a swap_remove-like implementation and added documentation. I cannot directly call swap_remove during `choose` because it makes the index unstable. The correct way is to start at the tail and swap tuples one by one.

yihozhang commented on src/scheduler.rs:
```diff
@@ -0,0 +1,449 @@
+        }
+
+        let mut rules = Vec::new();
+        let rulesets = std::mem::take(&mut self.rulesets);
+        collect_rules(ruleset, &rulesets, &mut rules);
+        let mut schedulers = std::mem::take(&mut self.schedulers);
```
I think this is because a mutable reference to self is used later during rule execution.
